url5e6a51063b_nuigroup_hci_for_java_tgz-pJ8-nui_squirt_demo_DemoSquirtApplicationJ8/src/nui/squirt/component/Slider.java

```diff
@@ -0,0 +1,204 @@
+import nui.squirt.Valuable;
+import nui.squirt.event.ValueEvent;
+import nui.squirt.listener.ValueListener;
+import processing.core.PApplet;
+import processing.core.PVector;
+
+
+public class Slider extends Rectangle implements Valuable {
+
+	private Collection<ValueListener> listeners = new ArrayList<ValueListener>();
+	
+	private float length;
+	
+	private float minValue;
+	private float maxValue;
+	private float value;
+	
+	private ControlPoint current = null;
+
+	public Slider(float x, float y, float l) {
+		super(x, y, 60, 90);
+		this.length = l;
+	}
+	
+	public Slider(float x, float y, float l, float minValue, float maxValue) {
+		this(x, y, l);
+		this.minValue = minValue;
+		this.maxValue = maxValue;
+		this.value = getCenterValue();
+	}
+	
+	public Slider(float x, float y, float l, float minValue, float maxValue, float initValue) {
+		this(x, y, l, minValue, maxValue);
+		this.value = initValue;
+	}
+
+	public void addValueListener(ValueListener l) {
+		listeners.add(l);
+	}
+
+	public void removeValueListener(ValueListener l) {
+		listeners.remove(l);
+	}
+
+	public void fireValueChanged(ValueEvent e) {
+		for (ValueListener l: listeners) {
+			l.valueChanged(e);
+		}
+	}
+	
+	public float getLength() {
+		return length;
+	}
+
+	public void setLength(float length) {
+		this.length = length;
+	}
+
+	public float getMinValue() {
+		return minValue;
+	}
+
+	public void setMinValue(float minValue) {
+		this.minValue = minValue;
+	}
+
+	public float getMaxValue() {
+		return maxValue;
+	}
+
+	public void setMaxValue(float maxValue) {
+		this.maxValue = maxValue;
+	}
+
+	public float getValue() {
+		return value;
+	}
+
+	public void setValue(float value) {
+		float oldValue = getValue();
+		
+		float v = value;
+		if (value > getMaxValue())
+			v = getMaxValue();
+		else if (value < getMinValue())
+			v = getMinValue();
+		this.value = v;
+		
+		AffineTransform t = getTransformMatrix();
+		double a = getValue()-oldValue;
+		double b = a/getValueRange();
+		double c = b*getLength();
+		double d = -c;
+		t.translate(0, d);
+		
+		ValueEvent e = new ValueEvent(this);
+		e.setOldValue(oldValue);
+		e.setNewValue(getValue());
+		fireValueChanged(e);
+	}
+
+	public float getCenterValue() {
+		return getValueRange()/2 + getMinValue();
+	}
+
+	public float getValueRange() {
+		return getMaxValue() - getMinValue();
+	}
+	
+	@Override
+	public void render(PApplet p) {
+		// Draw the path of the slider
+		p.stroke(25);
+		p.fill(75);
+
+		float h = (getValue()-getCenterValue()) / getValueRange() * getLength();
+		p.rect(0, h, 5, getLength());
+		
+		setStrokeColor(new Color(25));
+		super.render(p);
+		p.stroke(100);
+		p.line(-getWidth()/2, 0, getWidth()/2, 0);
+	}
+	
+	@Override
+	public boolean canAcceptMoreControlPoints() {
+		return current == null;
+	}
+	
+	@Override
+	public void controlPointCreated(ControlPoint cp) {
+		if (current == null) {
+			PVector l = transformToLocalSpace(new PVector(cp.getX(), cp.getY()));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;cp&#39; (index: 0 | type: ControlPoint). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other methods validating the parameter: &#39;controlPointUpdated&#39;, &#39;controlPointDied&#39;. The same parameter type is validated here for example: url5e6a51063b_nuigroup_hci_for_java_tgz-pJ8-nui_squirt_demo_DemoSquirtApplicationJ8/src/nui/squirt/component/Circle.java:107. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).